dace/transformation/dataflow/map_fusion.py

```diff
@@ -1,537 +1,1583 @@
+            raise ValueError("You specified both `only_inner_maps` and `only_toplevel_maps`.")
+
+        # Ensure that both have the same schedule
+        if map_entry_1.map.schedule != map_entry_2.map.schedule:
+            return False
+
+        # Fusing is only possible if the two entries are in the same scope.
+        scope = graph.scope_dict()
+        if scope[map_entry_1] != scope[map_entry_2]:
+            return False
+        elif self.only_inner_maps:
+            if scope[map_entry_1] is None:
+                return False
+        elif self.only_toplevel_maps:
+            if scope[map_entry_1] is not None:
+                return False
+
+        # We will now check if there exists a remapping that of the map parameter
+        if self.find_parameter_remapping(first_map=map_entry_1.map, second_map=map_entry_2.map) is None:
+            return False
 
-        # Success
         return True
 
-    def apply(self, graph: SDFGState, sdfg: SDFG):
+
+    def has_read_write_dependency(
+        self,
+        map_entry_1: nodes.MapEntry,
+        map_entry_2: nodes.MapEntry,
+        state: SDFGState,
+        sdfg: SDFG,
+    ) -> bool:
+        """Test if there is a read write dependency between the two maps to be fused.
+
+        The function checks two different things.
+        - The function will make sure that there is no read write dependency between
+            the input and output of the fused maps. For that it will inspect the
+            respective subsets.
+        - The second part partially checks the intermediate nodes, it mostly ensures
+            that there are not views and that they are not used as inputs or outputs
+            at the same time. However, the function will not check for read write
+            conflicts in this set, this is done in the partition function.
+
+        Returns:
+            `True` if there is a conflict between the maps that can not be handled.
+            If there is no conflict or if the conflict can be handled `False`
+            is returned.
+
+        Args:
+            map_entry_1: The entry node of the first map.
+            map_entry_2: The entry node of the second map.
+            state: The state on which we operate.
+            sdfg: The SDFG on which we operate.
         """
-            This method applies the mapfusion transformation.
-            Other than the removal of the second map entry node (SME), and the first
-            map exit (FME) node, it has the following side effects:
+        map_exit_1: nodes.MapExit = state.exit_node(map_entry_1)
+        map_exit_2: nodes.MapExit = state.exit_node(map_entry_2)
+
+        # Get the read and write sets of the different maps, note that Views
+        #  are not resolved yet.
+        access_sets: List[Dict[str, nodes.AccessNode]] = []
+        for scope_node in [map_entry_1, map_exit_1, map_entry_2, map_exit_2]:
+            access_set: Set[nodes.AccessNode] = self.get_access_set(scope_node, state)
```

Comment:
This only seems to collect access nodes which are connected as outputs to the map exits and inputs to the map entries, right? If so, are map-internal writes / reads taken care of?

Reply:
At first I had trouble understanding your question, because I thought you refers to the data that is used to exchange data between the maps, i.e. what `self.array` matches (which is not examined in this function, but in the partition function).
However, then I realized that you mean writes that are fully withing the map, i.e. the dataflow does not pass through the map entry/exit.   
Then you are right, this is not checked, because until now I have not realized that this might be an issue, since I assumed that every data has to pass through the maps.
However, now I realize that this is wrong (for example I know that `CompositeFusion` may put intermediates fully inside a map.

To handle this case I have made the following additions:
- Fusion is refused if both map bodies consist of an access node that refers to the same data.
- Fusion is refused if a map body consists of an access node that refers to global data (this restriction could be lifted).